File: libatsc3Module/src/main/java/org/ngbp/libatsc3/ndk/entities/held/HeldXmlParser.java
package org.ngbp.libatsc3.ndk.entities.held;

import android.util.Log;

import androidx.annotation.Nullable;

import org.ngbp.libatsc3.ndk.XmlUtils;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;

public class HeldXmlParser {
    private static final String ENTRY_HELD = "HTMLEntryPackage";

    @Nullable
    public Held parseXML(String xmlPayload) {
        try {
            XmlPullParser parser = XmlUtils.newParser(xmlPayload);

            parser.nextTag();
            parser.require(XmlPullParser.START_TAG, null, "HELD");
            while (parser.next() != XmlPullParser.END_TAG) {
                if (parser.getEventType() != XmlPullParser.START_TAG) {
                    continue;
                }

                String name = parser.getName();
                if (name.equals(ENTRY_HELD)) {
                    return readHeld(parser);
                } else {
                    XmlUtils.skip(parser);
                }
            }
        } catch (XmlPullParserException | IOException e) {
            Log.e("HeldXmlParser","exception in parsing: "+e);
        }

        return null;
    }

    private Held readHeld(XmlPullParser parser) throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, null, ENTRY_HELD);

        Held held = new Held();

        int attrCount = parser.getAttributeCount();
        for (int i = 0; i < attrCount; i++) {
            String attrName = parser.getAttributeName(i);
            switch (attrName) {
                case "appContextId":
                    held.appContextId = parser.getAttributeValue(i);
                    break;
                case "appRendering":
                    held.appRendering = XmlUtils.strToBool(parser.getAttributeValue(i));
                    break;
                case "bcastEntryPackageUrl":
                    held.bcastEntryPackageUrl = parser.getAttributeValue(i);
                    break;
                case "bcastEntryPageUrl":
                    held.bcastEntryPageUrl = parser.getAttributeValue(i);
                    break;
                case "coupledServices":
                    held.coupledServices = XmlUtils.strToInt(parser.getAttributeValue(i));
                    break;
                default:
                    break;
            }
        }

        return held;
    }
}

File: libatsc3Module/src/main/java/org/ngbp/libatsc3/entities/held/HeldXmlParser.kt
package org.ngbp.libatsc3.entities.held

import android.util.Log
import org.ngbp.libatsc3.utils.XmlUtils
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException

class HeldXmlParser {
    fun parseXML(xmlPayload: String): Atsc3Held? {
        val packages = arrayListOf<Atsc3HeldPackage>()
        try {
            val parser = XmlUtils.newParser(xmlPayload)

            parser.nextTag()
            parser.require(XmlPullParser.START_TAG, null, "HELD")

            while (parser.next() != XmlPullParser.END_DOCUMENT) {
                if (parser.eventType != XmlPullParser.START_TAG) {
                    continue
                }

                if (parser.name == ENTRY_HELD_PACKAGE) {
                    packages.add(readHeld(parser))
                } else {
                    XmlUtils.skip(parser)
                }
            }

            return Atsc3Held(packages)
        } catch (e: XmlPullParserException) {
            Log.e("HeldXmlParser", "exception in parsing: $e")
        } catch (e: IOException) {
            Log.e("HeldXmlParser", "exception in parsing: $e")
        }

        return null
    }

    @Throws(IOException::class, XmlPullParserException::class)
    private fun readHeld(parser: XmlPullParser): Atsc3HeldPackage {
        parser.require(XmlPullParser.START_TAG, null, ENTRY_HELD_PACKAGE)

        val pkg = Atsc3HeldPackage()

        val attrCount = parser.attributeCount
        for (i in 0 until attrCount) {
            val value = parser.getAttributeValue(i)
            when (parser.getAttributeName(i)) {
                "appContextId" -> pkg.appContextId = value
                "requiredCapabilities" -> pkg.requiredCapabilities = XmlUtils.strToLong(value)
                "appRendering" -> pkg.appRendering = XmlUtils.strToBool(value)
                "clearAppContextCacheDate" -> pkg.clearAppContextCacheDate = XmlUtils.strToDate(value)
                "bcastEntryPackageUrl" -> pkg.bcastEntryPackageUrl = value
                "bcastEntryPageUrl" -> pkg.bcastEntryPageUrl = value
                "bbandEntryPageUrl" -> pkg.bbandEntryPageUrl = value
                "validFrom" -> pkg.validFrom = XmlUtils.strToDate(value)
                "validUntil" -> pkg.validUntil = XmlUtils.strToDate(value)
                "coupledServices" -> pkg.coupledServices = XmlUtils.strToListOfInt(value)
                else -> {
                    // skip
                }
            }
        }

        return pkg
    }

    companion object {
        private const val ENTRY_HELD_PACKAGE = "HTMLEntryPackage"
    }
}